include/nanobind/eigen/sparse.h

```diff
@@ -0,0 +1,124 @@
+
+    Copyright (c) 2023 Wenzel Jakob
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include <nanobind/tensor.h>
+#include <nanobind/eigen/dense.h>
+#include <Eigen/SparseCore>
+
+static_assert(EIGEN_VERSION_AT_LEAST(3, 3, 0),
+              "Eigen matrix support in nanobind requires Eigen >= 3.3.0");
+
+NAMESPACE_BEGIN(NB_NAMESPACE)
+
+NAMESPACE_BEGIN(detail)
+
+template <typename Scalar, int Flags, typename StorageIndex>
+using EigenMapSparseMatrix = Eigen::Map<Eigen::SparseMatrix<Scalar, Flags, StorageIndex>>;
+
+/// Caster for Eigen::SparseMatrix<T>
+template <typename T> struct type_caster<T, enable_if_t<is_eigen_sparse_v<T>>> {
+    using Scalar = typename T::Scalar;
+    using StorageIndex = typename T::StorageIndex;
+    using Index = typename T::Index;
+    static constexpr bool row_major = T::IsRowMajor;
+
+    using ScalarTensor = tensor<Scalar, shape<any>, c_contig, device::cpu>;
+    using StorageIndexTensor = tensor<StorageIndex, shape<any>, c_contig, device::cpu>;
+
+    using ScalarCaster = make_caster<ScalarTensor>;
+    using StorageIndexCaster = make_caster<StorageIndexTensor>;
+
+    NB_TYPE_CASTER(T, const_name<row_major>("scipy.sparse.csr_matrix[",
+                                            "scipy.sparse.csc_matrix[")
+                   + make_caster<Scalar>::Name + const_name("]"));
+
+    ScalarCaster scalar_caster;
+    StorageIndexCaster storageindex_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+        auto obj = borrow<object>(src);
+        try {
+            object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+            if (!obj.type().is(matrix_type))
+                obj = matrix_type(obj);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        if (object data_o = obj.attr("data"); !scalar_caster.from_python(data_o, flags, cleanup))
+            return false;
+        ScalarTensor& values = scalar_caster.value;
+
+        if (object indices_o = obj.attr("indices"); !storageindex_caster.from_python(indices_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& inner_indices = storageindex_caster.value;
+
+        if (object indptr_o = obj.attr("indptr"); !scalar_caster.from_python(indptr_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& outer_indices = storageindex_caster.value;
+
+        object shape_o = obj.attr("shape"), nnz_o = obj.attr("nnz");
+        Index rows, cols, nnz;
+        try {
+            if (len(shape_o) != 2)
+                return false;
+            rows = cast<Index>(shape_o[0]);
+            cols = cast<Index>(shape_o[1]);
+            nnz = cast<Index>(nnz_o);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        value = EigenMapSparseMatrix<Scalar,
```

Comment:
The formatting is a bit unwieldy here. Perhaps you can define this type with a `using SparseMap = ..` declaration at the top of the caster and then reference it here?